File: code/rabbitmq/java/client/src/main/java/dev/learn/rabbitmq/L01.java
package dev.learn.rabbitmq;

import static dev.learn.rabbitmq.Broker.*;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import java.util.concurrent.CountDownLatch;

/** Lesson 1: the same producer and consumer as the C# ones, printing the same lines. */
final class L01 {

    private L01() {
    }

    static void send() throws Exception {
        try (Connection connection = connect("l01-send-java");
             Channel channel = connection.createChannel()) {
            var server = connection.getServerProperties();
            System.out.println("connected to " + server.get("product") + " " + server.get("version"));

            AMQP.Queue.DeclareOk queue = channel.queueDeclare("hello", true, false, false, null);
            System.out.println("queue " + queue.getQueue() + ": " + queue.getMessageCount() + " message(s) waiting");

            for (int i = 1; i <= 3; i++) {
                // exchange "", routing key = queue name, no properties
                channel.basicPublish("", "hello", null, bytes("Hello " + i));
                System.out.println("sent Hello " + i);
            }
        }
    }

    static void receive() throws Exception {
        try (Connection connection = connect("l01-receive-java");
             Channel channel = connection.createChannel()) {
            AMQP.Queue.DeclareOk queue = channel.queueDeclare("hello", true, false, false, null);
            System.out.println("queue " + queue.getQueue() + ": " + queue.getMessageCount() + " message(s) waiting");

            CountDownLatch waiting = new CountDownLatch(queue.getMessageCount());
            // The callback runs on the client's consumer thread pool, one message at a time per channel.
            channel.basicConsume("hello", true,
                    (consumerTag, delivery) -> {
                        System.out.println("received " + text(delivery.getBody()));
                        waiting.countDown();
                    },
                    consumerTag -> { });
            await(waiting, "the waiting messages");
        }
    }
}

File: code/rabbitmq/csharp/L01.cs
using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using static Learn.Rabbit.Broker;

namespace Learn.Rabbit;

// Lesson 1: a producer and a consumer that never meet. The broker keeps the messages in between.
static class L01
{
    public static async Task Send()
    {
        await using IConnection connection = await ConnectAsync("l01-send");
        Console.WriteLine($"connected to {Text(connection.ServerProperties?["product"])} {Text(connection.ServerProperties?["version"])}");

        // A channel is a lightweight session multiplexed on the connection; almost every AMQP operation happens on one.
        await using IChannel channel = await connection.CreateChannelAsync();

        // Declaring is idempotent: it creates the queue, or checks that the existing one has the same settings.
        QueueDeclareOk queue = await channel.QueueDeclareAsync("hello", durable: true, exclusive: false, autoDelete: false);
        Console.WriteLine($"queue {queue.QueueName}: {queue.MessageCount} message(s) waiting");

        for (var i = 1; i <= 3; i++)
        {
            // The default exchange "" delivers to the queue named by the routing key.
            await channel.BasicPublishAsync(exchange: "", routingKey: "hello", body: Bytes($"Hello {i}"));
            Console.WriteLine($"sent Hello {i}");
        }
    }

    public static async Task Receive()
    {
        await using IConnection connection = await ConnectAsync("l01-receive");
        await using IChannel channel = await connection.CreateChannelAsync();
        QueueDeclareOk queue = await channel.QueueDeclareAsync("hello", durable: true, exclusive: false, autoDelete: false);
        Console.WriteLine($"queue {queue.QueueName}: {queue.MessageCount} message(s) waiting");

        // The example stops once it has read what was waiting; a real consumer runs until the application stops.
        var remaining = (int)queue.MessageCount;
        var done = new TaskCompletionSource();
        if (remaining == 0)
        {
            done.SetResult();
        }

        var consumer = new AsyncEventingBasicConsumer(channel);
        consumer.ReceivedAsync += (_, delivery) =>
        {
            Console.WriteLine($"received {Text(delivery.Body)}");
            if (--remaining == 0)
            {
                done.TrySetResult();
            }
            return Task.CompletedTask;
        };

        // autoAck: true means the broker forgets a message as soon as it sends it. Lesson 4 shows what that costs.
        await channel.BasicConsumeAsync("hello", autoAck: true, consumer);
        await WaitAsync(done.Task, "the waiting messages");
    }
}
